spinalcordtoolbox/scripts/sct_analyze_lesion.py

```diff
@@ -531,11 +542,14 @@ def ___pve_weighted_avg(self, im_mask_data, im_atlas_data):
     def __keep_only_indices(self, image, indices):
         """Keep values defined by indices, and set all other coordinates to zero."""
         image_out = np.zeros_like(image)
-        image_out[indices] = image[indices]
+        if isinstance(indices, list) and len(indices) == 3:  # Lists of [x], [y], and [z] coords produced by np.where()
```

Comment:
But `np.where` returns a `tuple`, not a `list`?

I'm concerned that this might silently return the wrong result, because `image[indices]` and `image[:, :, indices]` both "work", but with extremely different results:
```py3
>>> data = np.arange(24).reshape((2, 3, 4)); data
array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

>>> data[np.where(data < 5)]
array([0, 1, 2, 3, 4])

>>> data[:, :, np.where(data < 5)]
array([[[[ 0,  0,  0,  0,  0],
         [ 0,  0,  0,  0,  1],
         [ 0,  1,  2,  3,  0]],

        [[ 4,  4,  4,  4,  4],
         [ 4,  4,  4,  4,  5],
         [ 4,  5,  6,  7,  4]],

        [[ 8,  8,  8,  8,  8],
         [ 8,  8,  8,  8,  9],
         [ 8,  9, 10, 11,  8]]],


       [[[12, 12, 12, 12, 12],
         [12, 12, 12, 12, 13],
         [12, 13, 14, 15, 12]],

        [[16, 16, 16, 16, 16],
         [16, 16, 16, 16, 17],
         [16, 17, 18, 19, 16]],

        [[20, 20, 20, 20, 20],
         [20, 20, 20, 20, 21],
         [20, 21, 22, 23, 20]]]])
```

And yes, the type is wrong:
```py3
>>> isinstance(np.where(data < 5), list)
False
>>> isinstance(np.where(data < 5), tuple)
True
```

Reply:
Re: list vs. tuple, thank you for catching this :hearts:

Re: `np.where` vs. `[:, :, slice]`... I was under the impression that, we're only using the indexing/slicing as a roundabout way to zero out all other voxels, the shape of the the intermediate array (1D vs. 3D) doesn't matter? But maybe I'm wrong here... :thinking:

Reply:
I guess it would be a lot clearer if `indices_to_keep` had a single form (e.g. tuple of x,y,z coordinates) rather than multiple forms (tuple *or* slice number). So, I can definitely refactor things to avoid this inconsistency.

Reply:
In the commits I added, I made it so that `indices_to_keep` is always something that can be used directly to index a numpy array. Hopefully this works? Let me know what you think.

Reply:
> Re: np.where vs. [:, :, slice]... I was under the impression that, because we're only using the indexing/slicing as a roundabout way to zero out all other voxels, the shape of the the intermediate array (1D vs. 3D) doesn't matter? But maybe I'm wrong here... 🤔

In my example above with `np.arange`, I think it would actually make a difference. In particular:
* Most of them would redundantly get set to zero several times (not a real problem).
* All the entries would get set to zero, not just the first five (yes a problem).

```py3
>>> import numpy as np
>>> data = np.arange(24).reshape((2, 3, 4)); data
array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

>>> data[np.where(data < 5)] = 0; data
array([[[ 0,  0,  0,  0],
        [ 0,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

>>> data[:, :, np.where(data < 5)] = 0; data
array([[[0, 0, 0, 0],
        [0, 0, 0, 0],
        [0, 0, 0, 0]],

       [[0, 0, 0, 0],
        [0, 0, 0, 0],
        [0, 0, 0, 0]]])
```

Reply:
I very much like the new structure you've added with `slice(None)` representing `:`, and I have no complaints! :tada:

(As an aside, I'm not 100% sure the `[:, :, np.where(data < 5)]` example given is 1:1 with the old code (since before, it was more like `[:, :, n_slice]`), but that's kind of moot now as you've undoubtedly improved `indices_to_keep`, so I have nothing much to add, hehe.)